genattr.py

```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/python3 --
+import os
+import re
+import subprocess
+import sys
+import tempfile
+import stat
+
+def main(argv, write_cmd, read_cmd, expected_stdout):
+    match_re = re.compile(r'\A[A-Za-z0-9._/+,-]+\Z')
+    write_cmd.write('set_current_time 1\n')
+    def bad(e):
+        raise e
+    def do_emit(path, context):
+        context = ':'.join(('system_u', 'object_r', context, 's0'))
+        write_cmd.write('ea_set /%s security.selinux %s\n' % (path, context))
+        read_cmd.write('ea_get /%s security.selinux\n' % (path,))
+        expected_stdout.write('''debugfs: ea_get /%s security.selinux
+security.selinux (%d) = "%s"
+
+''' % (path, len(context), context))
+    def emit(dirpath, name):
+        path = os.path.join(dirpath, name)
+        context = 'modules_object_t'
+        if path.startswith(argv + '/build'):
+            context = 'usr_t'
+        elif dirpath == argv:
+            if name.startswith('modules.'):
+                context = 'modules_dep_t'
+            elif name == 'build':
+                context = 'usr_t'
+        do_emit(path, context)
+    for i in ('initramfs', 'vmlinuz', 'firmware/' + argv):
+        assert stat.S_ISREG(os.lstat(joined).st_mode), 'build bug'
+        write_cmd.write(f'set_inode_field /{i} flags 16\n')
+        do_emit(i, 'modules_object_t')
+    do_emit('firmware', 'modules_object_t')
+    do_emit('lost+found', 'modules_object_t')
+    do_emit('', 'modules_object_t')
+    for dirpath, dirnames, filenames in os.walk(argv, onerror=bad, topdown=True):
+        for i in dirnames:
+            if not match_re.match(i):
+                raise ValueError("Invalid directory name %r" % os.path.join(dirpath, i))
+            emit(dirpath, i)
+        for i in filenames:
+            joined = os.path.join(dirpath, i)
+            if not match_re.match(i):
+                raise ValueError(f"Invalid file name {joined!r}")
+            if stat.S_ISREG(os.lstat(joined).st_mode):
+                write_cmd.write(f'set_inode_field {joined} flags 16\n')
```

Comment:
Why specifically? The modules.img is mounted read-only anyway.

Reply:
To prevent `restorecon` from mangling it even if the SELinux policy changes in the future.  See QubesOS/qubes-issues#5765.

Reply:
Does it actually help, even with overlayfs on top? If so, it can stay.